src/esp/nav/PathFinder.cpp

```diff
@@ -850,20 +850,43 @@ T PathFinder::tryStep(const T& start, const T& end) {
   static const int MAX_POLYS = 256;
   dtPolyRef polys[MAX_POLYS];
 
+  dtStatus startStatus, endStatus;
   dtPolyRef startRef, endRef;
   vec3f pathStart, pathEnd;
-  std::tie(std::ignore, startRef, pathStart) =
+  std::tie(startStatus, startRef, pathStart) =
       projectToPoly(start, navQuery_, filter_);
-  std::tie(std::ignore, endRef, pathEnd) =
-      projectToPoly(end, navQuery_, filter_);
+  std::tie(endStatus, endRef, pathEnd) = projectToPoly(end, navQuery_, filter_);
+
+  if (dtStatusFailed(startStatus) || dtStatusFailed(endStatus)) {
+    return start;
+  }
+
   vec3f endPoint;
   int numPolys;
   navQuery_->moveAlongSurface(startRef, pathStart.data(), pathEnd.data(),
                               filter_, endPoint.data(), polys, &numPolys,
                               MAX_POLYS);
 
-  if (numPolys == 0)
-    return {NAN, NAN, NAN};
+  // If there isn't any possible path between start and end, just return start,
+  // that is cleanest
+  if (numPolys == 0) {
+    return start;
+  }
+
+  // According to recast's code
+  // (https://github.com/recastnavigation/recastnavigation/blob/master/Detour/Source/DetourNavMeshQuery.cpp#L2006-L2007),
+  // the endPoint is not guaranteed to be actually on the surface of the
+  // navmesh, it seems to be in 99.9% of cases for us, but there are some
+  // extreme edge cases where it won't be, so explicitly get the height of the
+  // surface at the endPoint and set its height to that.
+  {
+    float endPointHeight;
+    // Note, this will never fail as endPoint is always within in the poly
+    // polys[numPolys - 1]
+    navQuery_->getPolyHeight(polys[numPolys - 1], endPoint.data(),
```

Comment:
Why not just pass &endPoint[1] here? You would avoid the temporary and the scope.

Reply:
I didn't want to dig through the code to make sure that is safe (it seems kinda scary to me).

Reply:
What is the concern exactly? I don't see the hazard.

However, on line 888, you are unconditionally updating endPoint[1]. If getPolyHeight were to somehow fail and not update endPointHeight, you would be copying uninitialized memory to endPoint[1]. Passing &endPoint[1], directly would avoid this issue.

Reply:
Looks like it's fine.  I generally prefer to not alias pointers.  There are optimizations the compiler can make if you don't alias pointers (although you need to use `__restrict__` to specify that they aren't aliased).

Reply:
Interesting, I knew about strict-aliasing and type-punning (not the case here) which can cause undefined behavior but was not aware of issues of non type-punned aliasing. But yeah, looks like if you use __restrict__ you can avoid a copy in certain cases. I don't think this is one of those cases though. The main benefit seems to be when there is a loop iterating over an array of data.

https://software.intel.com/en-us/articles/pointer-aliasing-and-vectorization
https://devblogs.nvidia.com/cuda-pro-tip-optimize-pointer-aliasing/